executor: enable application-defined environment variables

cmatsuoka opened the pull request:
Allow application-defined variables to be used during step scriptlet
execution. Applications can use the pre-step callback to set step
environment variables or the prologue callback to set global environment
variables.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----
CRAFT-1070

cmatsuoka commented:
> As a user of craft-parts, I'd appreciate some documentation with new features. Thanks for considering!

The callbacks API is still unstable, but we'll document it after it's considered stable for production use (we're exercising the experimental API with application-defined variables in Snapcraft and some changes may still be required).

sergiusens commented on tests/unit/executor/test_environment.py:
```diff
@@ -165,9 +170,11 @@ def test_generate_step_environment_no_project_name(new_dir):
         export PLUGIN_ENVVAR="from_plugin"
-        ## User Environment
+        ## Application environment
+        export APP_ENVVAR="from_app"
```
the test name does not match what is being tested

cmatsuoka replied:
Removed application variable setting from this test.

sergiusens reviewed (commented):
Looks good, but if we are not doing docs, an integration test would be nice. Docs should be fine if we document the api as unstable too

cmatsuoka commented:
Reworked the PR a bit to separate environment variables into global and step environment variables. This will allow global environment variable substitution later. Also added an integration that uses pre-step callbacks to set application-defined step environment variables and the prologue callback to set application-defined global environment variables.

sergiusens commented on craft_parts/executor/environment.py:
```diff
@@ -63,16 +59,22 @@ def generate_step_environment(
 
-        print("## User Environment", file=run_environment)
-        for env in user_build_environment:
+        print("## Application environment", file=run_environment)
```
application environment should come before part and plugin as these could reference the application provided variables

sergiusens commented on craft_parts/executor/environment.py:
```diff
@@ -88,7 +90,8 @@ def _basic_environment_for_part(part: Part, *, step_info: StepInfo) -> Dict[str,
     """
-    part_environment: Dict[str, str] = _get_step_environment(step_info)
+    part_environment: Dict[str, str] = _get_global_environment(step_info.project_info)
+    part_environment.update(_get_step_environment(step_info))
```
maybe update the part_environment dict with information from the application here as from the point of view of the user, it is all part of the same thing

cmatsuoka replied:
Updated to include global variables when getting the step environment. The separate call for global will be used later when expanding global variables.

cmatsuoka commented:
Addressed issues and rebased on current main.
